### PR TITLE
KAFKA-14588 ZK configuration moved to ZkConfig

### PR DESCRIPTION
**Pull request by nizhikov**
This PR move ZK related properties from scala `KafkaConfig` to java `ZkConfig`.
Similar to `TopicConfig` and other classes.

It's required, because `ConfigCommand` depends on `ZkConfig` and command itself will be moved to `tools` module.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: server-common/src/main/java/org/apache/kafka/server/metrics/ClientMetricsConfigs.java
Comment (mimaison) [inline server-common/src/main/java/org/apache/kafka/server/metrics/ClientMetricsConfigs.java:1]: These files don't seem to be related to ZooKeeper configs, why are we moving them in this PR?

File: core/src/main/scala/kafka/server/DynamicConfig.scala [hunk 1]
```diff
@@ -35,87 +33,22 @@ import scala.jdk.CollectionConverters._
 object DynamicConfig {
 
   object Broker {
-    // Properties
-    val LeaderReplicationThrottledRateProp = "leader.replication.throttled.rate"
```
Comment (mimaison) [inline core/src/main/scala/kafka/server/DynamicConfig.scala]: Do we have to move these fields in this PR? These are not ZooKeeper configs

**Review by mimaison — COMMENTED** [review]
Thanks for the PR. I left a few questions.

**Comment by nizhikov** [comment]
Hello @mimaison 

This PR based on #14871 and includes changes mad in it.
I think it will be better to review #14871 and after it returns to these changes.

Will add "[WIP]" lablel to this PR.
Sorry for any inconvinience.

**Comment by nizhikov** [comment]
Hello @mimaison @ijuma 

This PR is ready for review.
Please, take a look.

**Comment by nizhikov** [comment]
Hello @mimaison @ijuma

Can you, please, take a look?

**Comment by chia7712** [comment]
@nizhikov please rebase the code, and I will take a look later

**Comment by nizhikov** [comment]
@chia7712 Done.

Please, note, there are several other PR's with same goal #15569 #15575 which are close to mergeable state in my opinion.

File: server/src/main/java/org/apache/kafka/server/config/ZkConfig.java [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+    public static final boolean ZK_SSL_CLIENT_ENABLE = false;
+    public static final String ZK_SSL_PROTOCOL = "TLSv1.2";
+    public static final String ZK_SSL_ENDPOINT_IDENTIFICATION_ALGORITHM = "HTTPS";
+    public static final boolean ZK_SSL_CRL_ENABLE = false;
+    public static final boolean ZK_SSL_OCSP_ENABLE = false;
+
+    static {
+        Map<String, String> zkSslConfigToSystemPropertyMap = new HashMap<>();
+
+        zkSslConfigToSystemPropertyMap.put(ZK_SSL_CLIENT_ENABLE_PROP, ZKClientConfig.SECURE_CLIENT);
```
Comment (chia7712) [inline server/src/main/java/org/apache/kafka/server/config/ZkConfig.java]: We can use constants to replace `ZKClientConfig.SECURE_CLIENT` and `ZKClientConfig.ZOOKEEPER_CLIENT_CNXN_SOCKET`, and then `server` module can get rid of zk dependency.

File: server/src/main/java/org/apache/kafka/server/config/ZkConfig.java [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+        zkSslConfigToSystemPropertyMap.put(ZK_SSL_PROTOCOL_PROP, "zookeeper.ssl.protocol");
+        zkSslConfigToSystemPropertyMap.put(ZK_SSL_ENABLED_PROTOCOLS_PROP, "zookeeper.ssl.enabledProtocols");
+        zkSslConfigToSystemPropertyMap.put(ZK_SSL_CIPHER_SUITES_PROP, "zookeeper.ssl.ciphersuites");
+        zkSslConfigToSystemPropertyMap.put(ZK_SSL_ENDPOINT_IDENTIFICATION_ALGORITHM_PROP, "zookeeper.ssl.hostnameVerification");
+        zkSslConfigToSystemPropertyMap.put(ZK_SSL_CRL_ENABLE_PROP, "zookeeper.ssl.crl");
+        zkSslConfigToSystemPropertyMap.put(ZK_SSL_OCSP_ENABLE_PROP, "zookeeper.ssl.ocsp");
+
+        ZK_SSL_CONFIG_TO_SYSTEM_PROPERTY_MAP = Collections.unmodifiableMap(zkSslConfigToSystemPropertyMap);
+
+        ZK_SSL_CLIENT_ENABLE_DOC = "Set client to use TLS when connecting to ZooKeeper." +
```
Comment (chia7712) [inline server/src/main/java/org/apache/kafka/server/config/ZkConfig.java]: it should be fine to initialize `ZK_SSL_CLIENT_ENABLE_DOC` when declaring it (line#60)

File: build.gradle [hunk 4]
```diff
@@ -831,6 +831,17 @@ project(':server') {
     implementation libs.metrics
 
     implementation libs.slf4jApi
+    implementation(libs.zookeeper) {
```
Comment (chia7712) [inline build.gradle]: this can get removed (see my another comment)

File: checkstyle/import-control-jmh-benchmarks.xml [hunk 5]
```diff
@@ -58,4 +58,7 @@
 
     <subpackage name="cache">
     </subpackage>
+    <subpackage name="metadata">
```
Comment (chia7712) [inline checkstyle/import-control-jmh-benchmarks.xml]: it seems be unnecessary

File: checkstyle/import-control-server.xml [hunk 6]
```diff
@@ -82,6 +82,9 @@
       <allow class="org.apache.kafka.server.authorizer.AuthorizableRequestContext" />
       <allow pkg="org.apache.kafka.server.telemetry" />
     </subpackage>
+    <subpackage name="config">
+      <allow class="org.apache.zookeeper.client.ZKClientConfig" />
```
Comment (chia7712) [inline checkstyle/import-control-server.xml]: ditto. we can remove it

File: checkstyle/import-control.xml [hunk 7]
```diff
@@ -605,6 +605,8 @@
         <allow pkg="org.apache.kafka.metadata" />
         <allow pkg="org.eclipse.jetty.client"/>
         <allow class="org.apache.kafka.storage.internals.log.CleanerConfig" />
+        <allow class="org.apache.kafka.server.config.ZkConfig" />
+        <allow class="org.apache.zookeeper.client.ZKClientConfig" />
```
Comment (chia7712) [inline checkstyle/import-control.xml]: ditto

File: server/src/main/java/org/apache/kafka/server/config/ZkConfig.java [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+package org.apache.kafka.server.config;
+
+import org.apache.zookeeper.client.ZKClientConfig;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+public class ZkConfig {
```
Comment (chia7712) [inline server/src/main/java/org/apache/kafka/server/config/ZkConfig.java]: It seems we don't want to instantiate this class, so could you rename it to `ZkConfigs` and make it be a `final` class?

**Review by chia7712 — COMMENTED** [review]
@nizhikov thanks for handling this issue! a couple of comments left. PTAL

File: server/src/main/java/org/apache/kafka/server/config/ZkConfig.java (hunk 8 above)
Comment (nizhikov) [reply server/src/main/java/org/apache/kafka/server/config/ZkConfig.java]: Done.

File: checkstyle/import-control.xml (hunk 7 above)
Comment (nizhikov) [reply checkstyle/import-control.xml]: Done.

File: checkstyle/import-control-server.xml (hunk 6 above)
Comment (nizhikov) [reply checkstyle/import-control-server.xml]: Done.

File: checkstyle/import-control-jmh-benchmarks.xml (hunk 5 above)
Comment (nizhikov) [reply checkstyle/import-control-jmh-benchmarks.xml]: Done.

File: build.gradle (hunk 4 above)
Comment (nizhikov) [reply build.gradle]: Done.

File: server/src/main/java/org/apache/kafka/server/config/ZkConfig.java (hunk 3 above)
Comment (nizhikov) [reply server/src/main/java/org/apache/kafka/server/config/ZkConfig.java]: Done

File: server/src/main/java/org/apache/kafka/server/config/ZkConfig.java (hunk 2 above)
Comment (nizhikov) [reply server/src/main/java/org/apache/kafka/server/config/ZkConfig.java]: Done.

**Comment by nizhikov** [comment]
@chia7712 All your comments resolved. Please, take a look one more time.

**Comment by chia7712** [comment]
@nizhikov Could you please rebase it again? Sorry for late review :(

**Comment by nizhikov** [comment]
@chia7712 Conflicts resolved.

File: server/src/main/java/org/apache/kafka/server/config/ZkConfig.java (hunk 3 above)
Comment (nizhikov) [reply server/src/main/java/org/apache/kafka/server/config/ZkConfig.java]: I reverted those changes and leave my first version.
Because, java8 fails to load class if we use map values inside String constants definition.

https://ci-builds.apache.org/blue/organizations/jenkins/Kafka%2Fkafka-pr/detail/PR-15075/16/pipeline/9 

```
[2024-03-26T15:45:33.431Z] java.lang.StackOverflowError
[2024-03-26T15:45:33.431Z] 	at com.sun.tools.javac.comp.Attr.visitLambda(Attr.java:2296)
[2024-03-26T15:45:33.431Z] 	at com.sun.tools.javac.tree.JCTree$JCLambda.accept(JCTree.java:1624)
[2024-03-26T15:45:33.431Z] 	at com.sun.tools.javac.comp.Attr.attribTree(Attr.java:576)
[2024-03-26T15:45:33.431Z] 	at com.sun.tools.javac.comp.Attr.visitLambda(Attr.java:2435)
[2024-03-26T15:45:33.431Z] 	at com.sun.tools.javac.tree.JCTree$JCLambda.accept(JCTree.java:1624)
[2024-03-26T15:45:33.431Z] 	at com.sun.tools.javac.comp.Attr.attribTree(Attr.java:576)
[2024-03-26T15:45:33.431Z] 	at com.sun.tools.javac.comp.Attr.visitLambda(Attr.java:2435)
[2024-03-26T15:45:33.431Z] 	at com.sun.tools.javac.tree.JCTree$JCLambda.accept(JCTree.java:1624)
[2024-03-26T15:45:33.431Z] 	at com.sun.tools.javac.comp.Attr.attribTree(Attr.java:576)
[2024-03-26T15:45:33.431Z] 	at com.sun.tools.javac.comp.Attr.visitLambda(Attr.java:2435)
[2024-03-26T15:45:33.431Z] 	at com.sun.tools.javac.tree.JCTree$JCLambda.accept(JCTree.java:1624)
[2024-03-26T15:45:33.431Z] 	at com.sun.tools.javac.comp.Attr.attribTree(Attr.java:576)
[2024-03-26T15:45:33.431Z] 	at com.sun.tools.javac.comp.Attr.visitLambda(Attr.java:2435)
[2024-03-26T15:45:33.431Z] 	at com.sun.tools.javac.tree.JCTree$JCLambda.accept(JCTree.java:1624)
```

Comment (chia7712) [reply server/src/main/java/org/apache/kafka/server/config/ZkConfig.java]: oh, nice finding. please ignore my previous comment :)

**Comment by nizhikov** [comment]
@chia7712 CI are ready and seems OK for me.

**Comment by chia7712** [comment]
```
./gradlew cleanTest :tools:test --tests GetOffsetShellTest.testTopicPatternArgWithPartitionsArg :trogdor:test --tests CoordinatorTest.testTaskRequestWithOldStartMsGetsUpdated :connect:mirror:test --tests MirrorConnectorsIntegrationExactlyOnceTest.testReplicateSourceDefault --tests MirrorConnectorsWithCustomForwardingAdminIntegrationTest.testReplicateSourceDefault :core:test --tests ReplicaManagerTest.testRemoteFetchExpiresPerSecMetric --tests LogDirFailureTest.testIOExceptionDuringCheckpoint :clients:test --tests KafkaAdminClientTest.testClientSideTimeoutAfterFailureToReceiveResponse
```
failed tests can pass on my local, so I'm about to merge it

**Comment by OmniaGM** [comment]
@nizhikov it would be nice if we have a followup cleanup on this to change PROP suffix to CONFIG as this seems to be the pattern for others and am doing the same for the rest of KafkaConfig properties.

**Comment by nizhikov** [comment]
@OmniaGM Thanks for noticing. https://github.com/apache/kafka/pull/15614/files prepared
